File: src/cli/commands/diff.ts
/**
 * `homer diff` —— 漂移的详细差异（docs/m1-plan.md §1.7 / §2-P1-M4）。
 *
 * 冻结签名（§1.7）：
 *   DiffOptions / runDiff(opts): string   —— 渲染好的多行文本
 *
 * 输出约定（计划原文）：
 *   - merge 键行：`key: old → new`（缺失一侧渲染为 `(无)`）
 *   - mirror 文件：行级 `+` / `-`（LCS）
 *   - 空输出 = 无漂移（调用方据此判断；不打印「无漂移」占位，保证脚本可判空）
 *
 * `sources` 第二参数与 status 一致：单测 / M2 注入三方快照用；生产不传时走 store + 实时扫描。
 */

import { computeDrift, type AdapterDrift, type CategoryDrift, type MirrorOp } from '../../core/engine/index.js';
import type { AdapterSnapshot, SnapshotEntry, SnapshotFiles } from '../../core/types.js';
import { isPlainObject } from '../../core/entry-kind.js';
import { loadConfig } from '../../core/config.js';
import {
  CliError,
  collectSnapshotSources,
  diffLines,
  formatValue,
  resolveHomerPaths,
  sourceErrorMessages,
  type CliDriftSources,
} from '../render.js';

export interface DiffOptions {
  homerHome?: string;
  adapter?: string;
  category?: string;
}

export const DIFF_USAGE = `用法: homer diff [options]

显示漂移的详细差异：
  - merge 文件：按键行输出 \`key: old → new\`
  - mirror 文件：按行输出 \`+\` / \`-\`（LCS diff）

选项:
  --home <dir>       homer 工作区（默认 $HOMER_HOME 或 ~/.homer）
  --adapter <id>     只看某个 adapter
  --category <name>  只看某个分类
  -h, --help         显示本帮助

无漂移时输出为空（exit 0）。`;

/* ------------------------------------------------------------------ */
/* JSON 键级 diff（渲染用，取 old/new 值）                              */
/* ------------------------------------------------------------------ */

interface JsonSlot { has: boolean; value: unknown }

interface KeyRow { path: string; before: unknown; after: unknown }

/**
 * merge 键行的取值：仅当条目标记为 kind:'json' 时才尝试解析。
 * 解析失败（如 kind:'json' 但内容损坏）→ undefined，调用方据此把该文件降级为
 * mirror 行级 diff（与 drift.ts 的 isDegraded 对齐），不再走「宽松解析」的第三套逻辑。
 */
function slotOf(entry: SnapshotEntry | undefined): JsonSlot | undefined {
  if (entry === undefined) return { has: false, value: undefined };
  if (entry.kind !== 'json') return undefined;
  try {
    return { has: true, value: JSON.parse(entry.content) as unknown };
  } catch {
    return undefined;
  }
}

function isJsonEntry(entry: SnapshotEntry | undefined): boolean {
  return entry === undefined || entry.kind === 'json';
}

function childSlot(parent: JsonSlot, key: string): JsonSlot {
  if (!parent.has || !isPlainObject(parent.value)) return { has: false, value: undefined };
  // hasOwnProperty 守卫：`__proto__` / `constructor` 不得沿原型链取值（同 merge.ts childSlot）。
  if (!Object.prototype.hasOwnProperty.call(parent.value, key)) return { has: false, value: undefined };
  return { has: true, value: parent.value[key] };
}

function slotsEqual(a: JsonSlot, b: JsonSlot): boolean {
  if (a.has !== b.has) return false;
  if (!a.has) return true;
  return JSON.stringify(a.value) === JSON.stringify(b.value);
}

/**
 * 递归收集键变更（与引擎 diffJson 的粒度对齐：双方都是对象 → 递归到叶子；
 * 整体新增/删除的子树只记子树根键）。根值非对象时记 `$`。
 * 双侧 slot 均为 undefined 表示「该条目不可 JSON 解析」→ 整体跳过（走行级 diff）。
 */
function collectKeyDiffs(base: JsonSlot | undefined, local: JsonSlot | undefined, prefix: string, out: KeyRow[]): void {
  if (base === undefined || local === undefined) return;
  if (!base.has && !local.has) return;

  if (base.has && local.has && isPlainObject(base.value) && isPlainObject(local.value)) {
    const keys = new Set<string>([...Object.keys(base.value), ...Object.keys(local.value)]);
    for (const key of [...keys].sort()) {
      collectKeyDiffs(childSlot(base, key), childSlot(local, key), prefix === '' ? key : `${prefix}.${key}`, out);
    }
    return;
  }

  if (slotsEqual(base, local)) return;
  out.push({
    path: prefix === '' ? '$' : prefix,
    before: base.has ? base.value : undefined,
    after: local.has ? local.value : undefined,
  });
}

/* ------------------------------------------------------------------ */
/* 渲染                                                                */
/* ------------------------------------------------------------------ */

function findFiles(snapshots: readonly AdapterSnapshot[], adapterId: string, category: string): SnapshotFiles {
  return snapshots.find((s) => s.adapterId === adapterId)?.categories.find((c) => c.category === category)?.files
    ?? new Map<string, SnapshotEntry>();
}

/**
 * 渲染一个文件的差异（base→local 方向）。返回是否真的产出了行。
 * 第三方向（remote，仅当 remote !== base 时）以 `远端(↓)` 段落追加。
 */
function renderFileDiff(
  out: string[],
  relPath: string,
  base: SnapshotEntry | undefined,
  local: SnapshotEntry | undefined,
  remote: SnapshotEntry | undefined,
  conflict = false,
): boolean {
  const collect = (from: string, to: string): string[] => diffLines(from, to);

  let lines: string[] = [];
  if (base === undefined && local !== undefined) {
    lines = collect('', local.content);
  } else if (base !== undefined && local === undefined) {
    lines = collect(base.content, '');
  } else if (base !== undefined && local !== undefined && base.content !== local.content) {
    lines = collect(base.content, local.content);
  }

  let remoteLines: string[] = [];
  const remoteChanged = base === undefined ? remote !== undefined : remote === undefined || remote.content !== base.content;
  if (remoteChanged) {
    remoteLines = collect(base?.content ?? '', remote?.content ?? '');
  }

  if (lines.length === 0 && remoteLines.length === 0) return false;

  out.push(`  ${conflictMark(conflict)}${relPath}`);
  out.push(...lines);
  if (remoteLines.length > 0) {
    out.push('    远端(↓)');
    out.push(...remoteLines);
  }
  return true;
}

/** 冲突标记前缀（minor 3）：命中 mergeConflicts / conflict op 的行以 `⚡` 打头。 */
const CONFLICT_MARK = '⚡ ';

function conflictMark(conflict: boolean): string {
  return conflict ? CONFLICT_MARK : '';
}

function renderMirrorCategory(
  out: string[],
  adapterId: string,
  drift: CategoryDrift,
  base: SnapshotFiles,
  local: SnapshotFiles,
  remote: SnapshotFiles,
): void {
  let rendered = false;
  for (const op of drift.ops) {
    if (op.type === 'noop') continue;
    if (!rendered) {
      out.push(`${adapterId}/${drift.category}`);
      rendered = true;
    }
    renderFileDiff(out, op.path, base.get(op.path), local.get(op.path), remote.get(op.path), op.type === 'conflict');
  }
}

function renderMergeCategory(
  out: string[],
  adapterId: string,
  drift: CategoryDrift,
  base: SnapshotFiles,
  local: SnapshotFiles,
  remote: SnapshotFiles,
): void {
  // 降级文件（merge 分类里 kind:'file'）由 ops 接管，走行级 diff。
  const degraded = new Set(drift.ops.filter((op) => op.type !== 'noop').map((op) => op.path));
  // 冲突打标（minor 3）：mergeConflicts 的 keyPath 既可能是键点路径（'theme'），
  // 也可能是文件级路径（relPath，来自 base 缺失 / modify-vs-delete 等文件级分支）。
  const conflictKeys = new Set(drift.mergeConflicts.map((conflict) => conflict.keyPath));
  const conflictOpPaths = new Set(drift.ops.filter((op) => op.type === 'conflict').map((op) => op.path));
  const fileConflicts = (relPath: string): boolean => conflictKeys.has(relPath) || conflictOpPaths.has(relPath);
  const keyConflicts = (keyPath: string): boolean => conflictKeys.has(keyPath);

  let rendered = false;
  const head = (): void => {
    if (!rendered) {
      out.push(`${adapterId}/${drift.category}`);
      rendered = true;
    }
  };

  for (const op of drift.ops) {
    if (op.type === 'noop') continue;
    head();
    renderFileDiff(out, op.path, base.get(op.path), local.get(op.path), remote.get(op.path), op.type === 'conflict');
  }

  const paths = new Set<string>([...base.keys(), ...local.keys(), ...remote.keys()]);
  for (const relPath of [...paths].sort()) {
    if (degraded.has(relPath)) continue;

    const baseEntry = base.get(relPath);
    const localEntry = local.get(relPath);
    const remoteEntry = remote.get(relPath);
    if (!isJsonEntry(baseEntry) || !isJsonEntry(localEntry) || !isJsonEntry(remoteEntry)) continue;

    const rows: KeyRow[] = [];
    collectKeyDiffs(slotOf(baseEntry), slotOf(localEntry), '', rows);

    const remoteRows: KeyRow[] = [];
    if (remoteEntry?.content !== baseEntry?.content) {
      collectKeyDiffs(slotOf(baseEntry), slotOf(remoteEntry), '', remoteRows);
    }
    if (rows.length === 0 && remoteRows.length === 0) continue;

    head();
    out.push(`  ${conflictMark(fileConflicts(relPath))}${relPath}`);
    for (const row of rows) {
      out.push(`    ${conflictMark(keyConflicts(row.path))}${row.path}: ${formatValue(row.before)} → ${formatValue(row.after)}`);
    }
    for (const row of remoteRows) {
      out.push(`    ↓ ${conflictMark(keyConflicts(row.path))}${row.path}: ${formatValue(row.before)} → ${formatValue(row.after)}`);
    }
  }
}

function isEmptyDrift(drift: AdapterDrift): boolean {
  return drift.categories.every((c) => c.push === 0 && c.pull === 0 && c.conflicts === 0);
}

/**
 * 渲染漂移差异文本。无漂移 → 空字符串。
 * 无 homer.json → 抛 CliError（分发层提示先 init，exit 1）。
 *
 * 采集告警（M-A）以 `⚠ ...` 行置顶：diff 的空输出是脚本判据，
 * root 不可读时不能静默输出空（那会被误读为「无漂移」）。
 */
export function runDiff(opts: DiffOptions, sources?: CliDriftSources): string {
  const paths = resolveHomerPaths(opts.homerHome);
  const config = loadConfig(paths);
  if (config === undefined) {
    throw new CliError(
      `未找到 homer 配置: ${paths.configFile}`,
      '请先运行 `homer init` 生成 homer.json 与 store 快照。',
    );
  }

  const src: CliDriftSources = sources ?? collectSnapshotSources(paths, config);
  const drifts = computeDrift(src.base, src.local, src.remote);
  const remoteSnapshots = src.remote ?? src.base;

  const out: string[] = sourceErrorMessages(src.errors ?? []).map((message) => `⚠ ${message}`);
  for (const adapter of drifts) {
    if (opts.adapter !== undefined && adapter.adapterId !== opts.adapter) continue;
    if (isEmptyDrift(adapter)) continue;

    for (const category of adapter.categories) {
      if (opts.category !== undefined && category.category !== opts.category) continue;
      if (category.push === 0 && category.pull === 0 && category.conflicts === 0) continue;

      const base = findFiles(src.base, adapter.adapterId, category.category);
      const local = findFiles(src.local, adapter.adapterId, category.category);
      const remote = findFiles(remoteSnapshots, adapter.adapterId, category.category);

      if (category.mode === 'mirror') {
        renderMirrorCategory(out, adapter.adapterId, category, base, local, remote);
      } else {
        renderMergeCategory(out, adapter.adapterId, category, base, local, remote);
      }
    }
  }

  return out.join('\n');
}

File: internal/cli/commands/diff.go
package commands

import (
	"errors"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"

	"github.com/zzjcool/homer-cli/internal/core"
	"github.com/zzjcool/homer-cli/internal/engine"
	"github.com/zzjcool/homer-cli/internal/orderedjson"
)

// DiffOptions controls the read-only detailed diff command.
type DiffOptions struct {
	HomerHome string
	// Home is a compatibility alias for callers that use the CLI flag name.
	Home     string
	Adapter  string
	Category string
}

const DIFF_USAGE = "用法: homer diff [options]\n\n显示漂移的详细差异：\n  - merge 文件：按键行输出 `key: old → new`\n  - mirror 文件：按行输出 `+` / `-`（LCS diff）\n\n选项:\n  --home <dir>       homer 工作区（默认 $HOMER_HOME 或 ~/.homer）\n  --adapter <id>     只看某个 adapter\n  --category <name>  只看某个分类\n  -h, --help         显示本帮助\n\n无漂移时输出为空（exit 0）。"

// DiffUsage is the idiomatic alias for the archived constant spelling.
const DiffUsage = DIFF_USAGE

type jsonSlot struct {
	has   bool
	value orderedjson.Value
}

type keyRow struct {
	path      string
	before    orderedjson.Value
	after     orderedjson.Value
	beforeHas bool
	afterHas  bool
}

func entryPointer(files core.SnapshotFiles, path string) *core.SnapshotEntry {
	entry, ok := files[path]
	if !ok {
		return nil
	}
	copy := entry
	return &copy
}

func findFiles(snapshots []core.AdapterSnapshot, adapterID, category string) core.SnapshotFiles {
	for _, snapshot := range snapshots {
		if snapshot.AdapterID != adapterID {
			continue
		}
		for _, item := range snapshot.Categories {
			if item.Category == category {
				if item.Files == nil {
					return core.SnapshotFiles{}
				}
				return item.Files
			}
		}
	}
	return core.SnapshotFiles{}
}

// slotOf returns valid=false for a declared JSON entry whose bytes cannot be
// parsed.  The caller then follows the engine's degraded-file path instead of
// inventing a key-level diff for malformed content.
func slotOf(entry *core.SnapshotEntry) (*jsonSlot, bool) {
	if entry == nil {
		return &jsonSlot{has: false}, true
	}
	if entry.Kind != "json" {
		return nil, false
	}
	value, err := orderedjson.Parse([]byte(entry.Content))
	if err != nil {
		return nil, false
	}
	return &jsonSlot{has: true, value: value}, true
}

func isJSONEntry(entry *core.SnapshotEntry) bool {
	return entry == nil || entry.Kind == "json"
}

func objectValue(value orderedjson.Value) (*orderedjson.Object, bool) {
	object, ok := value.(*orderedjson.Object)
	if !ok || object == nil {
		return nil, false
	}
	return object, true
}

func objectKeys(object *orderedjson.Object) []string {
	if object == nil {
		return nil
	}
	seen := make(map[string]struct{}, len(object.Keys))
	keys := make([]string, 0, len(object.Keys))
	for _, key := range object.Keys {
		if _, ok := object.M[key]; !ok {
			continue
		}
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		keys = append(keys, key)
	}
	extra := make([]string, 0)
	for key := range object.M {
		if _, ok := seen[key]; !ok {
			extra = append(extra, key)
		}
	}
	sort.Strings(extra)
	return append(keys, extra...)
}

func childSlot(parent *jsonSlot, key string) *jsonSlot {
	if parent == nil || !parent.has {
		return &jsonSlot{has: false}
	}
	object, ok := objectValue(parent.value)
	if !ok {
		return &jsonSlot{has: false}
	}
	value, ok := object.M[key]
	if !ok {
		return &jsonSlot{has: false}
	}
	return &jsonSlot{has: true, value: value}
}

func slotsEqual(a, b *jsonSlot) bool {
	if a == nil || b == nil {
		return a == b
	}
	if a.has != b.has {
		return false
	}
	if !a.has {
		return true
	}
	return orderedjson.DeepEqual(a.value, b.value)
}

func collectKeyDiffs(base, local *jsonSlot, prefix string, out *[]keyRow) {
	if base == nil || local == nil || (!base.has && !local.has) {
		return
	}
	baseObject, baseIsObject := objectValue(base.value)
	localObject, localIsObject := objectValue(local.value)
	if base.has && local.has && baseIsObject && localIsObject {
		seen := make(map[string]struct{})
		keys := make([]string, 0, len(baseObject.M)+len(localObject.M))
		for _, key := range append(objectKeys(baseObject), objectKeys(localObject)...) {
			if _, ok := seen[key]; ok {
				continue
			}
			seen[key] = struct{}{}
			keys = append(keys, key)
		}
		sort.Strings(keys)
		for _, key := range keys {
			path := key
			if prefix != "" {
				path = prefix + "." + key
			}
			collectKeyDiffs(childSlot(base, key), childSlot(local, key), path, out)
		}
		return
	}
	if slotsEqual(base, local) {
		return
	}
	path := prefix
	if path == "" {
		path = "$"
	}
	row := keyRow{path: path, beforeHas: base.has, afterHas: local.has}
	if base.has {
		row.before = base.value
	}
	if local.has {
		row.after = local.value
	}
	*out = append(*out, row)
}

func formatValueMissing(has bool, value orderedjson.Value) string {
	if !has {
		return "(无)"
	}
	if text, ok := value.(string); ok {
		return text
	}
	return compactJSON(value)
}

// compactJSON is the renderer-only compact counterpart to orderedjson's
// pretty file serializer.  It intentionally keeps ordered object keys and
// delegates string/number escaping/formatting to orderedjson for JS parity.
func compactJSON(value orderedjson.Value) string {
	switch typed := value.(type) {
	case nil:
		return "null"
	case *orderedjson.Object:
		if typed == nil || len(typed.M) == 0 {
			return "{}"
		}
		keys := objectKeys(typed)
		parts := make([]string, 0, len(keys))
		for _, key := range keys {
			keyText := string(orderedjson.Serialize(key))
			parts = append(parts, keyText+":"+compactJSON(typed.M[key]))
		}
		return "{" + strings.Join(parts, ",") + "}"
	case []orderedjson.Value:
		parts := make([]string, len(typed))
		for i, item := range typed {
			parts[i] = compactJSON(item)
		}
		return "[" + strings.Join(parts, ",") + "]"
	case string:
		return string(orderedjson.Serialize(typed))
	case bool:
		return strconv.FormatBool(typed)
	default:
		return string(orderedjson.Serialize(typed))
	}
}

func splitLines(text string) []string {
	if text == "" {
		return []string{}
	}
	lines := strings.Split(text, "\n")
	if len(lines) > 0 && lines[len(lines)-1] == "" {
		lines = lines[:len(lines)-1]
	}
	return lines
}

// diffLines is an LCS line diff with the exact +/-/context prefixes used by
// the TypeScript renderer.
func diffLines(before, after string) []string {
	if before == after {
		return []string{}
	}
	a := splitLines(before)
	b := splitLines(after)
	dp := make([][]int, len(a)+1)
	for i := range dp {
		dp[i] = make([]int, len(b)+1)
	}
	for i := len(a) - 1; i >= 0; i-- {
		for j := len(b) - 1; j >= 0; j-- {
			if a[i] == b[j] {
				dp[i][j] = dp[i+1][j+1] + 1
			} else if dp[i+1][j] >= dp[i][j+1] {
				dp[i][j] = dp[i+1][j]
			} else {
				dp[i][j] = dp[i][j+1]
			}
		}
	}

	result := make([]string, 0, len(a)+len(b))
	i, j := 0, 0
	for i < len(a) && j < len(b) {
		switch {
		case a[i] == b[j]:
			result = append(result, " "+a[i])
			i++
			j++
		case dp[i+1][j] >= dp[i][j+1]:
			result = append(result, "-"+a[i])
			i++
		default:
			result = append(result, "+"+b[j])
			j++
		}
	}
	for i < len(a) {
		result = append(result, "-"+a[i])
		i++
	}
	for j < len(b) {
		result = append(result, "+"+b[j])
		j++
	}
	return result
}

func conflictMark(conflict bool) string {
	if conflict {
		return "⚡ "
	}
	return ""
}

func renderFileDiff(out *[]string, relPath string, base, local, remote *core.SnapshotEntry, conflict bool) bool {
	lines := []string{}
	switch {
	case base == nil && local != nil:
		lines = diffLines("", local.Content)
	case base != nil && local == nil:
		lines = diffLines(base.Content, "")
	case base != nil && local != nil && base.Content != local.Content:
		lines = diffLines(base.Content, local.Content)
	}

	remoteChanged := false
	if base == nil {
		remoteChanged = remote != nil
	} else {
		remoteChanged = remote == nil || remote.Content != base.Content
	}
	remoteLines := []string{}
	if remoteChanged {
		before := ""
		if base != nil {
			before = base.Content
		}
		after := ""
		if remote != nil {
			after = remote.Content
		}
		remoteLines = diffLines(before, after)
	}
	if len(lines) == 0 && len(remoteLines) == 0 {
		return false
	}
	*out = append(*out, "  "+conflictMark(conflict)+relPath)
	*out = append(*out, lines...)
	if len(remoteLines) > 0 {
		*out = append(*out, "    远端(↓)")
		*out = append(*out, remoteLines...)
	}
	return true
}

func renderMirrorCategory(out *[]string, adapterID string, drift engine.CategoryDrift, base, local, remote core.SnapshotFiles) {
	headerWritten := false
	for _, op := range drift.Ops {
		if op.Type == "noop" {
			continue
		}
		if !headerWritten {
			*out = append(*out, adapterID+"/"+drift.Category)
			headerWritten = true
		}
		renderFileDiff(out, op.Path, entryPointer(base, op.Path), entryPointer(local, op.Path), entryPointer(remote, op.Path), op.Type == "conflict")
	}
}

func renderMergeCategory(out *[]string, adapterID string, drift engine.CategoryDrift, base, local, remote core.SnapshotFiles) {
	degraded := make(map[string]struct{})
	for _, op := range drift.Ops {
		if op.Type != "noop" {
			degraded[op.Path] = struct{}{}
		}
	}
	conflictKeys := make(map[string]struct{}, len(drift.MergeConflicts))
	for _, conflict := range drift.MergeConflicts {
		conflictKeys[conflict.KeyPath] = struct{}{}
	}
	conflictOpPaths := make(map[string]struct{})
	for _, op := range drift.Ops {
		if op.Type == "conflict" {
			conflictOpPaths[op.Path] = struct{}{}
		}
	}
	fileConflict := func(path string) bool {
		_, key := conflictKeys[path]
		_, op := conflictOpPaths[path]
		return key || op
	}
	keyConflict := func(path string) bool {
		_, ok := conflictKeys[path]
		return ok
	}

	headerWritten := false
	head := func() {
		if !headerWritten {
			*out = append(*out, adapterID+"/"+drift.Category)
			headerWritten = true
		}
	}
	for _, op := range drift.Ops {
		if op.Type == "noop" {
			continue
		}
		head()
		renderFileDiff(out, op.Path, entryPointer(base, op.Path), entryPointer(local, op.Path), entryPointer(remote, op.Path), op.Type == "conflict")
	}

	paths := make(map[string]struct{}, len(base)+len(local)+len(remote))
	for path := range base {
		paths[path] = struct{}{}
	}
	for path := range local {
		paths[path] = struct{}{}
	}
	for path := range remote {
		paths[path] = struct{}{}
	}
	orderedPaths := make([]string, 0, len(paths))
	for path := range paths {
		orderedPaths = append(orderedPaths, path)
	}
	sort.Strings(orderedPaths)
	for _, relPath := range orderedPaths {
		if _, ok := degraded[relPath]; ok {
			continue
		}
		baseEntry := entryPointer(base, relPath)
		localEntry := entryPointer(local, relPath)
		remoteEntry := entryPointer(remote, relPath)
		if !isJSONEntry(baseEntry) || !isJSONEntry(localEntry) || !isJSONEntry(remoteEntry) {
			continue
		}
		baseSlot, baseOK := slotOf(baseEntry)
		localSlot, localOK := slotOf(localEntry)
		remoteSlot, remoteOK := slotOf(remoteEntry)
		if !baseOK || !localOK || !remoteOK {
			continue
		}
		rows := make([]keyRow, 0)
		collectKeyDiffs(baseSlot, localSlot, "", &rows)
		remoteRows := make([]keyRow, 0)
		remoteChanged := false
		if baseEntry == nil {
			remoteChanged = remoteEntry != nil
		} else {
			remoteChanged = remoteEntry == nil || remoteEntry.Content != baseEntry.Content
		}
		if remoteChanged {
			collectKeyDiffs(baseSlot, remoteSlot, "", &remoteRows)
		}
		if len(rows) == 0 && len(remoteRows) == 0 {
			continue
		}
		head()
		*out = append(*out, "  "+conflictMark(fileConflict(relPath))+relPath)
		for _, row := range rows {
			*out = append(*out, "    "+conflictMark(keyConflict(row.path))+row.path+": "+formatValueMissing(row.beforeHas, row.before)+" → "+formatValueMissing(row.afterHas, row.after))
		}
		for _, row := range remoteRows {
			*out = append(*out, "    ↓ "+conflictMark(keyConflict(row.path))+row.path+": "+formatValueMissing(row.beforeHas, row.before)+" → "+formatValueMissing(row.afterHas, row.after))
		}
	}
}

func emptyDrift(drift engine.CategoryDrift) bool {
	return drift.Push == 0 && drift.Pull == 0 && drift.Conflicts == 0
}

// RunDiff returns the detailed diff text.  Empty text is the successful
// no-drift result; callers should not add a placeholder line.
func RunDiff(opts DiffOptions, injected ...DriftSources) (string, error) {
	homerHome := opts.HomerHome
	if homerHome == "" {
		homerHome = opts.Home
	}
	paths := core.GetHomerPaths(func(key string) string {
		if key == "HOMER_HOME" && homerHome != "" {
			return homerHome
		}
		return os.Getenv(key)
	})
	config, err := core.LoadConfig(paths)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return "", core.NewCliError(fmt.Sprintf("未找到 homer 配置: %s；请先运行 `homer init` 生成 homer.json 与 store 快照。", paths.ConfigFile))
		}
		return "", err
	}

	var sources DriftSources
	if provided, ok := sourceFromArgs(injected); ok {
		sources = provided
	} else {
		sources, err = CollectSnapshotSources(paths, config)
		if err != nil {
			return "", err
		}
	}

	drifts := engine.ComputeDrift(sources.Base, sources.Local, sources.Remote)
	remoteSnapshots := sources.Remote
	if remoteSnapshots == nil {
		remoteSnapshots = sources.Base
	}
	out := make([]string, 0)
	errorMessages := append([]string(nil), sources.Errors...)
	errorMessages = append(errorMessages, sources.ErrorMessages...)
	errorMessages = append(errorMessages, sourceErrorMessages(sources.ScanErrors)...)
	for _, message := range errorMessages {
		out = append(out, "⚠ "+message)
	}
	for _, warning := range sources.Warnings {
		out = append(out, "⚠ "+warning)
	}

	for _, drift := range drifts {
		if opts.Adapter != "" && opts.Adapter != drift.AdapterID {
			continue
		}
		if emptyDrift(drift) {
			continue
		}
		if opts.Category != "" && opts.Category != drift.Category {
			continue
		}
		base := findFiles(sources.Base, drift.AdapterID, drift.Category)
		local := findFiles(sources.Local, drift.AdapterID, drift.Category)
		remote := findFiles(remoteSnapshots, drift.AdapterID, drift.Category)
		if drift.Mode == core.SyncModeMirror {
			renderMirrorCategory(&out, drift.AdapterID, drift, base, local, remote)
		} else {
			renderMergeCategory(&out, drift.AdapterID, drift, base, local, remote)
		}
	}
	return strings.Join(out, "\n"), nil
}

func runDiff(opts DiffOptions, injected ...DriftSources) (string, error) {
	return RunDiff(opts, injected...)
}
